browserless/client.ts

```diff
@@ -0,0 +1,973 @@
+    expires?: number;
+    httpOnly?: boolean;
+    secure?: boolean;
+  }>;
+  authenticate?: {
+    username: string;
+    password: string;
+  };
+  userAgent?: string;
+  emulatedFormFactor?: "mobile" | "desktop" | "none";
+  output?: "json" | "html";
+  gotoOptions?: {
+    timeout?: number;
+    waitUntil?: "load" | "domcontentloaded" | "networkidle0" | "networkidle2";
+  };
+  viewport?: {
+    width: number;
+    height: number;
+    deviceScaleFactor?: number;
+    isMobile?: boolean;
+    hasTouch?: boolean;
+    isLandscape?: boolean;
+  };
+}
+
+/**
+ * @description Parameters for download API
+ */
+export interface DownloadParams {
+  url: string;
+  gotoOptions?: {
+    timeout?: number;
+    waitUntil?: "load" | "domcontentloaded" | "networkidle0" | "networkidle2";
+  };
+  cookies?: Array<{
+    name: string;
+    value: string;
+    domain: string;
+    path?: string;
+    expires?: number;
+    httpOnly?: boolean;
+    secure?: boolean;
+  }>;
+  authenticate?: {
+    username: string;
+    password: string;
+  };
+  selector?: string;
+  timeout?: number;
+  rejectResourceTypes?: string[];
+  rejectRequestPattern?: string[];
+  userAgent?: string;
+}
+
+/**
+ * @description Client for interacting with Browserless APIs
+ */
+export class BrowserlessClient {
+  private token?: string;
+  private baseUrl: string;
+
+  constructor(
+    token?: string,
+    baseUrl: string = BASE_URL,
+  ) {
+    this.token = token;
+    this.baseUrl = baseUrl;
+  }
+
+  /**
+   * @description Convert blob to base64
+   */
+  private blobToBase64(blob: Blob): Promise<string> {
+    return new Promise((resolve, reject) => {
+      const reader = new FileReader();
+      reader.onloadend = () => {
+        // Remove data URL prefix
+        const base64 = (reader.result as string).split(",")[1];
+        resolve(base64);
+      };
+      reader.onerror = reject;
+      reader.readAsDataURL(blob);
+    });
+  }
+
+  /**
+   * @description Get HTML content from a URL
+   */
+  async getContent(params: ContentParams): Promise<{ content: string }> {
+    const url = new URL(`${this.baseUrl}/content`);
+    if (this.token) {
+      url.searchParams.append("token", this.token);
+    }
+
+    const response = await fetch(url.toString(), {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify(params),
+    });
+
+    if (!response.ok) {
+      let errorMessage =
+        `Browserless API failed: ${response.status} ${response.statusText}`;
+      try {
+        const errorData = await response.json();
+        errorMessage += `. Details: ${JSON.stringify(errorData)}`;
+      } catch (_) {
+        // Ignore JSON parsing error
+      }
+      throw new Error(errorMessage);
+    }
+
+    const content = await response.text();
+    return { content };
+  }
+
+  /**
+   * @description Download files using the browser
+   */
+  async download(params: DownloadParams): Promise<{ base64Image: string }> {
```

Comment:
it seems that all functions have the same "body":
given an URL, extract token, fetch, test for error, process response.

Maybe you should create a base function that receives the URL, and a lambda to process response, and wrap all the functions from this client.